firmware/tests/test_linksys.py

```diff
@@ -0,0 +1,178 @@
+                                        Ver.1.1.0
+                                        <br>
+                                        Datum der letzten Version: 02/01/2018
+                                        <br>
+                                        <a href="http://downloads.linksys.com/downloads/firmware/FW_EA6300_2.03.21.exe">Herunterladen</a>
+                                    </div>
+                                </div>
+                            </div>
+                        </body>
+                    </html>
+                 '''
+
+SEARCH_TEXT = '<h3>Firmware (für USA)</h3>Ver.1.203.23 (build 20394)<br>Datum der letzten Version: 08/23/2019<br>' \
+              '<a href="http://downloads.linksys.com/downloads/firmware/FW_EA6300_1.203.23.20394_prod.gpg.img">' \
+              'Herunterladen</a>'
+
+
+@pytest.fixture(scope='session', autouse=True)
+def spider_instance():
+    return linksys.LinksysSpider()
+
+
+@pytest.fixture(scope='function', autouse=True)
+def mocked_request(monkeypatch):
+    monkeypatch.setattr(linksys, 'Request', MockRequest)
+
+
+@pytest.mark.parametrize('response, expected', [(MockResponse(url='https://www.linksys.com/de/support/sitemap/', body=PRODUCT_LIST_PAGE),
+                                                 [('https://www.linksys.com/de/support-product?pid=1234', dict(device_name='Broadband Router')),
+                                                  ('https://www.linksys.com/de/support-product?pid=5678', dict(device_name='Modem Router'))])])
+def test_parse(spider_instance, response, expected):
+    for anchor, request in enumerate(list(spider_instance.parse(response=response))):
+        assert request.url == expected[anchor][0]
+        assert request.cb_kwargs == expected[anchor][1]
+
+
+@pytest.mark.parametrize('response, device_name, expected', [(MockResponse(url='https://www.linksys.com/de/support-product?pid=1234', body=PRODUCT_PAGE), 'Broadband Router', ('https://www.linksys.com/de/support-article?articleNum=1234', dict(device_name='Broadband Router')))])
+def test_parse_product(spider_instance, response, device_name, expected):
+    for request in spider_instance.parse_product(response=response, device_name=device_name):
+        assert request.url == expected[0]
+        assert request.cb_kwargs == expected[1]
+
+
+@pytest.mark.parametrize('response, device_name, expected', [(MockResponse(url='https://www.linksys.com/de/support-article?articleNum=1234', body=FIRMWARE_PAGE), 'Broadband Router', 1)])
+def test_parse_versions(monkeypatch, spider_instance, response, device_name, expected):
+    def mock_parse():
+        yield 1
+    with monkeypatch.context() as monkey:
+        monkey.setattr(linksys.LinksysSpider, 'parse_firmware', lambda *_, **__: mock_parse())
+        assert list(spider_instance.parse_versions(response=response, device_name=device_name))[0] == expected
+
+
+@pytest.mark.parametrize('device_name, version, expected', [('Broadband Router', FIRMWARE_PAGE,
```

Comment:
Formatting looks off. Just linebreak after `[` and continue with 4 spaces offset